src/utils/browser-connection-error-hints.ts

```diff
@@ -0,0 +1,39 @@
+import { flattenDeep, uniq } from 'lodash';
+import humanizeDuration from 'humanize-duration';
+import { BrowserSetOptions } from '../runner/interfaces';
+import BrowserConnection from '../browser/connection';
+import { LOCAL_BROWSER_INIT_TIMEOUT, REMOTE_BROWSER_INIT_TIMEOUT } from './browser-connection-timeouts';
+import { CONNECTION_ERROR_HINTS } from '../errors/types';
+import TEMPLATES from '../errors/runtime/templates';
+import renderTemplate from './render-template';
+
+// NOTE: hint about too high concurrency factor will be added to the error after exceeding this value
+const CONCURRENCY_FACTOR_UPPERBOUND = 3;
+
+function getUsedTimeoutMsg (browserInitTimeout?: number): string {
+    if (browserInitTimeout) {
+        const browserInitTimeoutStr = humanizeDuration(browserInitTimeout);
+
+        return `${browserInitTimeoutStr} for all browsers`;
+    }
+
+    const localInitTimeoutStr  = humanizeDuration(LOCAL_BROWSER_INIT_TIMEOUT);
+    const remoteInitTimeoutStr = humanizeDuration(REMOTE_BROWSER_INIT_TIMEOUT);
+
+    return `${localInitTimeoutStr} for local browsers and ${remoteInitTimeoutStr} for remotes`;
+}
+
+export function createHints (connections: BrowserConnection[], opts: BrowserSetOptions): string[] {
+    const warningsFromConnections  = uniq(flattenDeep(connections.map(bc => bc.warningLog.messages)));
+    const warningsFromBootstrapper = opts.warningLog.messages;
+
+    const hints = [ ...warningsFromConnections, ...warningsFromBootstrapper ];
+
+    if (opts.concurrency > CONCURRENCY_FACTOR_UPPERBOUND)
+        hints.push(renderTemplate(TEMPLATES[CONNECTION_ERROR_HINTS.tooHighConcurrencyFactor], opts.concurrency));
+
+    hints.push(renderTemplate(TEMPLATES[CONNECTION_ERROR_HINTS.useBrowserInitOption], getUsedTimeoutMsg(opts.browserInitTimeout)));
+    hints.push(renderTemplate(TEMPLATES[CONNECTION_ERROR_HINTS.restErrorCauses]));
+
+    return hints;
```

Comment:
Attach to this PR screenshots of terminal, to show how these hints looks in real world.

Reply:
I think screenshots too verbose and huge to be placed in a PR body, so I post them here. 

The compilation process took too long:
![tooLongCompilation](https://user-images.githubusercontent.com/26363017/103098722-e3cfb100-461c-11eb-88b1-57acdd8e100e.PNG)

"--browser-init-timeout" was specified, but container has too few resources available:
![tooLongCompilationWithCustomTimeout](https://user-images.githubusercontent.com/26363017/103098719-e29e8400-461c-11eb-8baf-1de377e10675.PNG)

Initialization time of the requested 7 instances of the Chromium browser took too long:
![chromiumBadPerformance](https://user-images.githubusercontent.com/26363017/103098720-e3371a80-461c-11eb-95a8-95e41166f5af.PNG)

Firefox didn't connected during the default timeout:
![defaultTimeout](https://user-images.githubusercontent.com/26363017/103098721-e3371a80-461c-11eb-95b0-d8a9eef6fd5f.png)